Sigils

stellanera98 opened the pull request:
- costs and effect are defined in datapack
- additional effect kinds can be registered in code
- there is only one sigil item, putting the sigil_type data component on it (ResourceKey<SigilType>) will determine its behaviour
- an item model for the sigil has to be placed at `assets/<type modid>/models/item/sigil_<type path>`
- translation keys `item.bloodmagic.sigil.<type path>` (item name) and `tooltip.bloodmagic.sigil.<type path>` are pretty much required
- if the sigil is a toggleable one (like fast miner or green growth) it needs the sigil_active data component set. if its not set the sigil will not toggle it on shift + use
- the sigil_active data component is also used as an item model override returning 1 for true and 0 for false, to be used on toggleable sigils


@dphaldes

Commit:
sigils

Commit:
it'd help to actually put the overrides in. also sigil_effect -> sigil_type for the type registry

dphaldes commented:
nitpick: 
> data/bloodmagic/bloodmagic/sigil_type

I wonder if this path be renamed to just "sigils" type doesnt really add much

dphaldes commented on src/main/java/wayoftime/bloodmagic/api/sigil/SigilEffect.java:
```diff
@@ -0,0 +1,68 @@
+    Supplier<MapCodec<ApplyPotionEffect>> APPLY_POTION_EFFECT = SIGIL_EFFECT_TYPE.register("apply_potion_effect", () -> ApplyPotionEffect.CODEC);
+    Supplier<MapCodec<DivinationEffect>> DIVINATION_EFFECT = SIGIL_EFFECT_TYPE.register("divination_effect", () -> DivinationEffect.CODEC);
+    Supplier<MapCodec<FluidPlaceEffect>> FLUID_PLACE_EFFECT = SIGIL_EFFECT_TYPE.register("fluid_place_effect", () -> FluidPlaceEffect.CODEC);
+    Supplier<MapCodec<FluidRemoveEffect>> FLUID_REMOVE_EFFECT = SIGIL_EFFECT_TYPE.register("fluid_remove_effect", () -> FluidRemoveEffect.CODEC);
```
Extract all the implementations and the map into a separate class in common instead of API.
keep the interface bare with only Sigil related content

stellanera98 replied:
so the effect and registry registering goes in common/sigil/SigilEffects and the interface itself stays in API. Codec definition stays in interface or would you put that with the other stuff?

dphaldes replied:
Codec can stay there. Extract the implementations and registry out

dphaldes commented on src/main/java/wayoftime/bloodmagic/api/sigil/SigilType.java:
```diff
@@ -0,0 +1,14 @@
+    public static final Codec<SigilType> CODEC = RecordCodecBuilder.create(builder -> builder.group(
+            Codec.INT.fieldOf("click_air_cost").forGetter(SigilType::airCost),
+            Codec.INT.fieldOf("click_block_cost").forGetter(SigilType::blockCost),
+            Codec.INT.fieldOf("click_entity_cost").forGetter(SigilType::entityCost),
```
I would call them "use_empty_cost", "use_block_cost" and "use_entity_cost" respectively

stellanera98 replied:
imo empty is weird there but use seems better than click

dphaldes commented:
> if the sigil is a toggleable one (like fast miner or green growth) it needs the sigil_active data component set. if its not set the sigil will not toggle it on shift + use

This should be done with the datapack directly. Add a field to the sigil types something along the line of "activatable" and use that to add and remove ACTIVE datacomponent on use. That way new activateable sigils are purely data pack driven and wont need the additional component set on them by default

dphaldes commented on src/main/java/wayoftime/bloodmagic/common/creativetab/BMTabs.java:
```diff
@@ -53,6 +59,25 @@ public class BMTabs {
+            ItemStack tmp = sigilStack.copy();
+            tmp.set(BMDataComponents.SIGIL_TYPE, key);
+            tmp.set(BMDataComponents.SIGIL_ACTIVE, false);
+            tab.accept(tmp);
```
not needed if you use my suggestion

dphaldes commented on src/main/java/wayoftime/bloodmagic/common/datacomponent/BMDataComponents.java:
```diff
@@ -29,6 +32,9 @@ public class BMDataComponents {
     public static final DeferredHolder<DataComponentType<?>, DataComponentType<Double>> DEMON_WILL_AMOUNT = DATA_COMPONENTS.registerComponentType("will_amount", builder -> builder.persistent(Codec.DOUBLE).networkSynchronized(ByteBufCodecs.DOUBLE));
     public static final DeferredHolder<DataComponentType<?>, DataComponentType<EnumWillType>> DEMON_WILL_TYPE = DATA_COMPONENTS.registerComponentType("will_type", builder -> builder.persistent(EnumWillType.CODEC).networkSynchronized(EnumWillType.STREAM_CODEC));
 
+    public static final DeferredHolder<DataComponentType<?>, DataComponentType<Boolean>> SIGIL_ACTIVE = DATA_COMPONENTS.registerComponentType("sigil_active", builder -> builder.persistent(Codec.BOOL).networkSynchronized(ByteBufCodecs.BOOL));
```
Can be made in a Unit type if you use my suggestion

dphaldes reviewed (approved):
Implementation looks solid. some small changes with names and suggestion

stellanera98 commented:
putting the activatable in datapack seems better, Ill get to that

Commit:
requested changes

# Conflicts:
#	src/main/java/wayoftime/bloodmagic/common/registry/BMRegistries.java

stellanera98 commented:
using Unit for the active type makes checks on if its active shorter, but toggling it turns from setting !state to an if/else which seems a bit excessive. I do like the boolean approach more but its not a big difference

dphaldes commented on src/main/java/wayoftime/bloodmagic/common/item/SigilItem.java:
Are you sure we want this file in API?

stellanera98 replied:
It shouldn't. When I moved it I was still mentally with the implementation version that required it to make new sigils, but thats not how it works anymore

dphaldes reviewed (approved):
Looks great!

Commit:
changes requested

stellanera98 commented:
- moved SigilItem back to common/items
- removed SigilType
- SigilEffect methods now return an int, this is then used as the cost
- SigilEffect now has `String getActiveTooltip(boolean isActive)` which by default returns translation keys for "Activated" and "Deactivated" but could be used to say something else where appropriate (see discord thread about it regarding the Exchange Sigil)
  - there has to be a better name for this


@dphaldes

dphaldes commented on src/main/java/wayoftime/bloodmagic/common/sigil/ApplyPotionEffect.java:
```diff
@@ -0,0 +1,61 @@
+            Codec.INT.fieldOf("upkeep").forGetter(ApplyPotionEffect::upkeep),
+            Codec.INT.optionalFieldOf("amplifier_others").forGetter(ApplyPotionEffect::otherAmplifier),
+            Codec.INT.optionalFieldOf("duration_others").forGetter(ApplyPotionEffect::otherTicks),
+            Codec.INT.optionalFieldOf("cost_others").forGetter(ApplyPotionEffect::otherCost)
```
What do these fields do? the keys in the datapack are no immediately clear

stellanera98 replied:
Nothing atm, they are intended for the beacon like effect with arrays

stellanera98 replied:
To specify, in 1.20 code the not fully implemented beacon like array effect with this sigil would have given the minecraft:haste effect for 600 ticks and an amplifier of 2 (so Haste III ingame) for 600 ticks for 1 HP damage (100 essence by default conversion) dealt to all the players in its AoE that receive it. I included those so it'd be easier later. See also SigilEffect#hasArrayEffect() and SigilEffect#arrayTick(...)

dphaldes commented on src/main/java/wayoftime/bloodmagic/common/sigil/ApplyPotionEffect.java:
```diff
@@ -0,0 +1,61 @@
+
+public record ApplyPotionEffect(Holder<MobEffect> effect, int selfAmplifier, int selfTicks, int upkeep, Optional<Integer> otherAmplifier, Optional<Integer> otherTicks, Optional<Integer> otherCost) implements SigilEffect {
+    public static final MapCodec<ApplyPotionEffect> CODEC = RecordCodecBuilder.mapCodec(builder -> builder.group(
+            MobEffect.CODEC.fieldOf("effect_type").forGetter(ApplyPotionEffect::effect),
```
I would call this "potion_effect_type" or "mob_effect_type" explicitly so that we dont get confused by Sigil Effects or any other effect

dphaldes commented on src/main/java/wayoftime/bloodmagic/common/sigil/ApplyPotionEffect.java:
```diff
@@ -0,0 +1,61 @@
+            MobEffect.CODEC.fieldOf("effect_type").forGetter(ApplyPotionEffect::effect),
+            Codec.INT.fieldOf("amplifier").forGetter(ApplyPotionEffect::selfAmplifier),
+            Codec.INT.fieldOf("duration").forGetter(ApplyPotionEffect::selfTicks),
+            Codec.INT.fieldOf("upkeep").forGetter(ApplyPotionEffect::upkeep),
```
if "upkeep" is a cost, lets add "_cost" suffix to it just in case

dphaldes commented on src/main/java/wayoftime/bloodmagic/common/sigil/FluidPlaceEffect.java:
```diff
@@ -0,0 +1,50 @@
+    public static final MapCodec<FluidPlaceEffect> CODEC = RecordCodecBuilder.mapCodec(builder -> builder.group(
+            BuiltInRegistries.FLUID.holderByNameCodec().fieldOf("fluid_type").forGetter(FluidPlaceEffect::fluid),
+            Codec.INT.fieldOf("tank_fill_amount").forGetter(FluidPlaceEffect::tankFillAmount),
+            Codec.INT.fieldOf("success_cost").forGetter(FluidPlaceEffect::cost)
```
can be just "cost", success is redundant

dphaldes commented on src/main/java/wayoftime/bloodmagic/common/sigil/FluidPlaceEffect.java:
```diff
@@ -0,0 +1,50 @@
+public record FluidPlaceEffect(Holder<Fluid> fluid, int tankFillAmount, int cost) implements SigilEffect {
+    public static final MapCodec<FluidPlaceEffect> CODEC = RecordCodecBuilder.mapCodec(builder -> builder.group(
+            BuiltInRegistries.FLUID.holderByNameCodec().fieldOf("fluid_type").forGetter(FluidPlaceEffect::fluid),
+            Codec.INT.fieldOf("tank_fill_amount").forGetter(FluidPlaceEffect::tankFillAmount),
```
should probably call it just "fluid_amount"  and use the same term in FluidRemoveEffect too

Commit:
removing array relatedness, changing other called out things
